core-api/src/main/java/com/optimizely/ab/OptimizelyUserContext.java

```diff
@@ -0,0 +1,186 @@
+ *    limitations under the License.
+ */
+package com.optimizely.ab;
+
+import com.optimizely.ab.optimizelydecision.OptimizelyDecideOption;
+import com.optimizely.ab.optimizelydecision.OptimizelyDecision;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class OptimizelyUserContext {
+    @Nonnull
+    private final String userId;
+
+    @Nonnull
+    private final Map<String, Object> attributes;
+
+    @Nonnull
+    private final Optimizely optimizely;
+
+    private static final Logger logger = LoggerFactory.getLogger(OptimizelyUserContext.class);
+
+    public OptimizelyUserContext(@Nonnull Optimizely optimizely,
+                                 @Nonnull String userId,
+                                 @Nonnull Map<String, ?> attributes) {
+        this.optimizely = optimizely;
+        this.userId = userId;
+        this.attributes = new ConcurrentHashMap<>(attributes);
+    }
+
+    public OptimizelyUserContext(@Nonnull Optimizely optimizely, @Nonnull String userId) {
+        this(optimizely, userId, new HashMap<>());
+    }
+
+    public String getUserId() {
+        return userId;
+    }
+
+    public Map<String, Object> getAttributes() {
+        return new HashMap<String, Object>(attributes);
+    }
+
+    public Optimizely getOptimizely() {
+        return optimizely;
+    }
+
+    /**
+     * Set an attribute for a given key.
+     *
+     * @param key An attribute key
+     * @param value An attribute value
+     */
+    public void setAttribute(@Nonnull String key, @Nonnull Object value) {
+        attributes.put(key, value);
+    }
+
+    /**
+     * Returns a decision result ({@link OptimizelyDecision}) for a given flag key and a user context, which contains all data required to deliver the flag.
+     * <ul>
+     * <li>If the SDK finds an error, it’ll return a decision with <b>null</b> for <b>variationKey</b>. The decision will include an error message in <b>reasons</b>.
+     * </ul>
+     * @param key A flag key for which a decision will be made.
+     * @param options A list of options for decision-making.
+     * @return A decision result.
+     */
+    public OptimizelyDecision decide(@Nonnull String key,
+                                     @Nonnull List<OptimizelyDecideOption> options) {
+        return optimizely.decide(this, key, options);
+    }
+
+    /**
+     * Returns a decision result ({@link OptimizelyDecision}) for a given flag key and a user context, which contains all data required to deliver the flag.
+     *
+     * @param key A flag key for which a decision will be made.
+     * @return A decision result.
+     */
+    public OptimizelyDecision decide(String key) {
+        return optimizely.decide(this, key, Collections.emptyList());
```

Comment:
change it to 
`this.decide(key, Collections.emptyList())`

Reply:
Will keep it to save one-step indirection.

Reply:
nit, but I agree with @msohailhussain . best to mitigate divergence within a class first. Also since Optimizely#Decide is private it doesn't need these overloaded signatures and they are purely a convenience on _this_ class.